src/t_hash.c

```diff
@@ -86,7 +88,12 @@ void hashTypeFreeVolatileSet(robj *o) {
 }
 
 void hashTypeTrackEntry(robj *o, void *entry) {
-    vset *set = hashTypeGetOrcreateVolatileSet(o);
+    vset *set;
+    if (hashTypeHasVolatileElements(o)) {
+        set = hashTypeGetVolatileSet(o);
+    } else {
+        set = hashTypeGetOrcreateVolatileSet(o);
+    }
     serverAssert(vsetAddEntry(set, entryGetExpiry, entry));
```

Comment:
Please do not include functional code inside an assert statement.  Use 2 lines.

Readers expect assertions to be validations only.  Assume that you should be able to blindly delete all assertions and the code will continue to function normally.